### PR TITLE
python310Packages.{jaxtyping,equinox}: update

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

Update `jaxtyping` and `equinox`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GaetanLepage** [comment]
I have the following problem when building `python310Packages.jaxtyping.passthru.tests`:
```
Finished executing pythonRemoveTestsDir
running install tests
no Makefile or custom installCheckPhase, doing nothing
pythonCatchConflictsPhase
/nix/store/q2zpnf017gccs9w9mb0ikdqhkf0k5v82-catch_conflicts.py:1: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
Found duplicated packages in closure for dependency 'jaxtyping':
  jaxtyping 0.2.23 (/nix/store/kmmqy0r284j5k7db70dinjxrx44kxmsq-python3.10-jaxtyping-0.2.23/lib/python3.10/site-packages)
  jaxtyping 0.2.23 (/nix/store/4l4jcgl3idx2hry5y5i72i16fhspdlir-python3.10-jaxtyping-0.2.23/lib/python3.10/site-packages)

Package duplicates found in closure, see above. Usually this happens if two packages depend on different version of the same dependency.
```

File: pkgs/development/python-modules/equinox/default.nix [hunk 1]
```diff
@@ -35,9 +36,17 @@ buildPythonPackage rec {
...
+    optax
     pytestCheckHook
   ];
 
+  # Those tests are failing upstream too.
+  # Relevant issue: https://github.com/patrick-kidger/equinox/issues/551
+  pytestFlagsArray = [
+    "--deselect=tests/test_nn.py::test_layer_norm"
+    "--deselect=tests/test_while_loop.py::test_nested_loops"
+  ];
```
Comment (samuela) [inline pkgs/development/python-modules/equinox/default.nix:48]: why not use `disabledTests = [...];`?

Comment (GaetanLepage) [reply pkgs/development/python-modules/equinox/default.nix]: You are right. Done !

**Review by samuela — APPROVED** [review]
diff LGTM

**Comment by GaetanLepage** [comment]
> diff LGTM

Unfortunately, the tests of `jaxtyping fail because of a collision between two "variation" of the `jaxtyping` derivation ending in the python environment.
I don't get why it starts to pause a problem now...

**Comment by samuela** [comment]
Hmm quick thoughts:
* Has anyone else packaged their own jaxtyping as a vendored part of a larger package?
* Are there multiple instantiations of jaxtyping in the transitive closure of the test dependencies? This could happen by eg one depending on jaxlib with `cudaSupport = true` and another with `cudaSupport = false`.

**Comment by GaetanLepage** [comment]
It seems that it is the fault of equinox. Updating equinox alone breaks `jaxtyping.tests`.

**Comment by GaetanLepage** [comment]
The only change in the `equinox` derivation besides the version bump is the addition of `optax` in the `nativeCheckInputs`.
However, removing it (as well as the tests that depend on it) doesn't solve our duplication problem.

**Comment by GaetanLepage** [comment]
Haha it is actually broken on master as well !
You can then merge this PR @samuela: it doesn't worsen the situation.

**Comment by samuela** [comment]
roger that, done! thanks @GaetanLepage !

**Pull request merged**